hibernate-core/src/main/java/org/hibernate/type/OneToOneType.java

```diff
@@ -159,6 +154,12 @@ public Serializable disassemble(Object value, SharedSessionContractImplementor s
 
 	@Override
 	public Object assemble(Serializable oid, SharedSessionContractImplementor session, Object owner) throws HibernateException {
+		if ( oid == null ) {
+			if ( uniqueKeyPropertyName != null ) {
+				return resolve( session.getContextEntityIdentifier( owner ), session, owner );
+			}
+			return null;
+		}
```

Comment:
@dreab8 , I saw there were quite some back and forth (see https://hibernate.atlassian.net/browse/HHH-15044); but I think after we revert back the dirty flush related code changes, the above code seems to become necessary again. I am not sure about this, though. Feel free to take over and work on top of my PR or even close it.

Reply:
Now there are quite some broken testing cases, but I believe it is due to the above code. Maybe it needs more polishing.

Reply:
Thanks. I agreed with Gavin King that most of the time when we touched OneToOne, we are making mistakes. I thought the original code changes by @deAtog introduced some bug in terms of misunderstanding of 'dirtiness' (without touching the dirtiness logic, his testing case still works). What puzzled me is the above new testing case added later. Seems the testing case is invalid one, but it seems to help to prove something.